server-spi/src/main/java/org/keycloak/models/LoginFailureProvider.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models;
+
+import org.keycloak.provider.Provider;
+
+/**
+ * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
+ */
+public interface LoginFailureProvider extends Provider {
```

Comment:
Login failures can be separated from user sessions because they don't have much in common. They each use a separate storage. MapUserSessionProvider can be then cleaner. 

Regarding the moving to `server-spi-private`, I am not sure. The new `LoginFailureProvider` is added to keycloak session which is in `server-spi`. I would need to add the `server-spi-private` as a dependency to `server-spi`.